### ISSUE TITLE
Add Zembretti Algorithm

### ISSUE THREAD

**Issue by GlennGoddard** [briis/hass-weatherflow2mqtt#133]
### New Feature

[https://github.com/17BIT0216/ALGORITHMS/blob/master/Zembrettin%20Algorithm/src/com/PrimesPackage/Main.java](https://github.com/17BIT0216/ALGORITHMS/blob/master/Zembrettin%20Algorithm/src/com/PrimesPackage/Main.java)

Simplified weather prediction based mainly on changes in atmospheric pressure.

This will rely on pressure trends that are currently partially available.  May need 3, 6, and 24 hour trends along with magnitude of change.

**Comment by leonardpitzu** [comment]
I've implemented this in JS and use some csv files to build the history data needed. I must say that although primitive it works quite well.

**Comment by GlennGoddard** [comment]
Are you using the package above or something else?  If it is something else, could you either post the code or link to it.  The more comparisons on how to do it (in my opinion) the better.

**Comment by leonardpitzu** [comment]
[betel_cast.txt](https://github.com/briis/hass-weatherflow2mqtt/files/7779421/betel_cast.txt)

No idea how i got it but here it is. It works pretty much ok - i've tested it quite a bit over the years and the delivered results are pretty accurate.

**Comment by GlennGoddard** [comment]
Just posting the full code here to be easier for copy and past translation to python.

```// beteljuice.com - near enough Zambretti Algorhithm 
// June 2008 - v1.0
// tweak added so decision # can be output

/* Negretti and Zambras 'slide rule' is supposed to be better than 90% accurate 
for a local forecast upto 12 hrs, it is most accurate in the temperate zones and about 09:00  hrs local solar time.
I hope I have been able to 'tweak it' a little better;-)  

This code is free to use and redistribute as long as NO CHARGE is EVER made for its use or output
*/

// ---- 'environment' variables ------------
var z_where       = 1;    // Northern = 1 or Southern = 2 hemisphere
var z_baro_top    = 1050;  // upper limits of your local 'weather window' (1050.0 hPa for UK)
var z_baro_bottom = 950;  // lower limits of your local 'weather window' (950.0 hPa for UK)

// usage:   forecast = betel_cast( z_hpa, z_month, z_wind, z_trend [, z_where] [, z_baro_top] [, z_baro_bottom])[0];

// z_hpa is Sea Level Adjusted (Relative) barometer in hPa or mB
// z_month is current month as a number between 1 to 12
// z_wind is English windrose cardinal eg. N, NNW, NW etc.
// NB. if calm a 'nonsense' value should be sent as z_wind (direction) eg. 1 or calm !
// z_trend is barometer trend: 0 = no change, 1= rise, 2 = fall
// z_where - OPTIONAL for posting with form
// z_baro_top - OPTIONAL for posting with form
// z_baro_bottom - OPTIONAL for posting with form
// [0] a short forecast text is returned
// [1] zambretti severity number (0 - 25) is returned ie. betel_cast() returns a two deep array


var z_forecast = new Array("Settled fine", "Fine weather", "Becoming fine", "Fine, becoming less settled", "Fine, possible showers", "Fairly fine, improving", "Fairly fine, possible showers early", "Fairly fine, showery later", "Showery early, improving", "Changeable, mending", "Fairly fine, showers likely", "Rather unsettled clearing later", "Unsettled, probably improving", "Showery, bright intervals", "Showery, becoming less settled", "Changeable, some rain", "Unsettled, short fine intervals", "Unsettled, rain later", "Unsettled, some rain", "Mostly very unsettled", "Occasional rain, worsening", "Rain at times, very unsettled", "Rain at frequent intervals", "Rain, very unsettled", "Stormy, may improve", "Stormy, much rain"); 

// equivalents of Zambretti 'dial window' letters A - Z
var rise_options   = new Array(25,25,25,24,24,19,16,12,11,9,8,6,5,2,1,1,0,0,0,0,0,0);
var steady_options = new Array(25,25,25,25,25,25,23,23,22,18,15,13,10,4,1,1,0,0,0,0,0,0);
var fall_options   = new Array(25,25,25,25,25,25,25,25,23,23,21,20,17,14,7,3,1,1,1,0,0,0);

var z_test = new Array();

// ---- MAIN FUNCTION --------------------------------------------------
function betel_cast( z_hpa, z_month, z_wind, z_trend, z_hemisphere, z_upper, z_lower) {
  if(z_hemisphere) z_where  = z_hemisphere; // used by input form
  if(z_upper) z_baro_top    = z_upper;      // used by input form
  if(z_lower) z_baro_bottom = z_lower;      // used by input form
  
  z_range    = z_baro_top - z_baro_bottom;
  z_constant = (z_range / 22).toFixed(3);

  z_season = (z_month >= 4 && z_month <= 9);   // true if 'Summer'
  if (z_where == 1) { // northern hemisphere
    if (z_wind == "N")        z_hpa += 6 / 100 * z_range;  
    else if (z_wind == "NNE") z_hpa += 5 / 100 * z_range;  
    else if (z_wind == "NE")  z_hpa += 5 / 100 * z_range;  
    else if (z_wind == "ENE") z_hpa += 2 / 100 * z_range;  
    else if (z_wind == "E")   z_hpa -= 0.5 / 100 * z_range;  
    else if (z_wind == "ESE") z_hpa -= 2 / 100 * z_range;  
    else if (z_wind == "SE")  z_hpa -= 5 / 100 * z_range;  
    else if (z_wind == "SSE") z_hpa -= 8.5 / 100 * z_range;  
    else if (z_wind == "S")   z_hpa -= 12 / 100 * z_range;  
    else if (z_wind == "SSW") z_hpa -= 10 / 100 * z_range;
    else if (z_wind == "SW")  z_hpa -= 6 / 100 * z_range;  
    else if (z_wind == "WSW") z_hpa -= 4.5 / 100 * z_range;
    else if (z_wind == "W")   z_hpa -= 3 / 100 * z_range;  
    else if (z_wind == "WNW") z_hpa -= 0.5 / 100 * z_range;  
    else if (z_wind == "NW")  z_hpa += 1.5 / 100 * z_range;  
    else if (z_wind == "NNW") z_hpa += 3 / 100 * z_range;
    
    if (z_season == 1) { // if Summer
      if (z_trend == 1)      z_hpa += 7 / 100 * z_range; // rising
      else if (z_trend == 2) z_hpa -= 7 / 100 * z_range; // falling
    } 
  } else { // southern hemisphere
    if (z_wind == "S")        z_hpa += 6 / 100 * z_range;  
    else if (z_wind == "SSW") z_hpa += 5 / 100 * z_range;  
    else if (z_wind == "SW")  z_hpa += 5 / 100 * z_range;  
    else if (z_wind == "WSW") z_hpa += 2 / 100 * z_range;  
    else if (z_wind == "W")   z_hpa -= 0.5 / 100 * z_range;  
    else if (z_wind == "WNW") z_hpa -= 2 / 100 * z_range;  
    else if (z_wind == "NW")  z_hpa -= 5 / 100 * z_range;  
    else if (z_wind == "NNW") z_hpa -= 8.5 / 100 * z_range;  
    else if (z_wind == "N")   z_hpa -= 12 / 100 * z_range;  
    else if (z_wind == "NNE") z_hpa -= 10 / 100 * z_range;
    else if (z_wind == "NE")  z_hpa -= 6 / 100 * z_range;  
    else if (z_wind == "ENE") z_hpa -= 4.5 / 100 * z_range;
    else if (z_wind == "E")   z_hpa -= 3 / 100 * z_range;  
    else if (z_wind == "ESE") z_hpa -= 0.5 / 100 * z_range;  
    else if (z_wind == "SE")  z_hpa += 1.5 / 100 * z_range;  
    else if (z_wind == "SSE") z_hpa += 3 / 100 * z_range;  
    
    if (z_season == 0) { // if Winter
      if (z_trend == 1)      z_hpa += 7 / 100 * z_range; // rising
      else if (z_trend == 2) z_hpa -= 7 / 100 * z_range; // falling
    } 
  } // END North / South

  if(z_hpa == z_baro_top) z_hpa = z_baro_top - 1;
  z_option = Math.floor((z_hpa - z_baro_bottom) / z_constant); 
  
  z_output = "";
  if(z_option < 0) {
    z_option = 0;
    z_output = "Exceptional Weather, ";
  }
  
  if(z_option > 21) {
    z_option = 21;
    z_output = "Exceptional Weather, ";
  }

  if (z_trend == 1) {   // rising
    z_output += z_forecast[rise_options[z_option]]; 
    z_test[1] = rise_options[z_option];
  } else if (z_trend == 2) {   // falling
    z_output += z_forecast[fall_options[z_option]]; 
    z_test[1] = fall_options[z_option];
  } else {   // must be 'steady'
    z_output += z_forecast[steady_options[z_option]]; 
    z_test[1] = steady_options[z_option];
  } 
  
  z_test[0] = z_output;
  
  return z_test; 
} // END function
```

**Comment by zacs** [comment]
It already exists in Python! https://github.com/jim-easterbrook/pywws/blob/master/src/pywws/forecast.py

It uses the PyWWS file access libraries, but those should be pretty easy to swap over. The only properties not relying on the "now" state are the barometric pressure.

**Comment by GlennGoddard** [comment]
Thanks, it can be hard to track down some of this information.
Glad to see you continue here from the HA community discussion we had (yes, I know you already were using briis other repositories).

**Comment by GlennGoddard** [comment]
Simple zambretti implementation
[https://github.com/sassoftware/iot-zambretti-weather-forcasting](https://github.com/sassoftware/iot-zambretti-weather-forcasting)

**Comment by GlennGoddard** [comment]
Note:
Reverse wind direction for Southern Hemisphere; need to verify other components but that appears the only change.
Will need Latitude for weather prediction already, so we can determine North or South from that information.
Will also need 3 hour trend and amount of change for pressure at sea level.

I am currently setting up a Node-Red Zambretti Forecast for testing.

**Comment by GlennGoddard** [comment]
For translations at least:
[https://github.com/3KUdelta/Solar_WiFi_Weather_Station/blob/master/Translation.h](https://github.com/3KUdelta/Solar_WiFi_Weather_Station/blob/master/Translation.h)

simple description:
[https://web.archive.org/web/20110610213848/http://www.meteormetrics.com/zambretti.htm](https://web.archive.org/web/20110610213848/http://www.meteormetrics.com/zambretti.htm)

Beteljuices Zambretti:
[http://www.beteljuice.co.uk/zambretti/forecast.html](http://www.beteljuice.co.uk/zambretti/forecast.html)
Building off this one for now.
Setting the High and Low pressure for an area from the stored All Time Max and Min Pressure

**Comment by GlennGoddard** [comment]
I have a test Zambretti running in Node-Red and output to Home Assistant.  I will run the test for some time before a PR.

I'm not sure what the best move for the forecast is.
Since the Zambretti is best when done at 0900 for an outcome 12 hours from then.
Should I have it run once a day at 0900?
Should I have it run once an hour and put it in an array to display forecast from +1 hour to +12 hours from current time?
or..Should I have it run continuously and average over an hour for the +11 hour forecast and array it to cover +1 to +11 hours?

**Comment by GlennGoddard** [comment]
Using the trend from here does not work perfectly for Zambretti.  Zambretti (after further review) only accepts stable if no rise or fall of 1.6 inHG in the last 3 hours.  Similar requirements for considering rise or fall.

I'm thinking of a 3 hr timer  that starts at 'stable' and is reset at a fall  or rise of 1.6 over the span of time.  This would require storing an array of pressures to compare to.

I'm open to ideas on how to accomplish this.

**Comment by GlennGoddard** [comment]
I have Zambretti working in Python and returns Zambretti Text and Number.  This is based on Beteljuice's Zambretti work.
Waiting on my previous PR to be accepted, which adds some prerequisites (briis might have previous PR in this weekend).

I still have some figuring out on the variables for the function call, like how to retrieve stored hi and lo press from inside helpers.py but one step at a time.

**Comment by GlennGoddard** [comment]
@briis The Zambretti as written returns two variables.
Do you want me to split it up to two separate functions?
I think it would probably be easier if it was two separate functions, so I will probably do that unless I hear otherwise.

**Comment by briis** [comment]
If the two variables should end up in two different sensors with a state value, then do two functions.

**Comment by GlennGoddard** [comment]
I put it in two functions, it was straight forward.
Tomorrow I should have the PR for Zambretti.  Since this got restructured, I don't know where the functions are actually called from anymore, so I'll have a PR for you tomorrow in the same state as the last one if that is ok?

**Comment by briis** [comment]
Very good Glenn. Just add the functions and sensor descriptions and I will take care of the rest. I will wait for this, and once completed, this will then become 3.1.0

**Comment by GlennGoddard** [comment]
#154 PR submitted for Zambretti
Not a perfect PR, I did not know how to get the All Time Sea Pressure: Max and Min as a variable inside the program.  (Might be environment variables)

**Comment by GlennGoddard** [comment]
I closed PR #154 and am adding translations

**Comment by GlennGoddard** [comment]
PR #155 adds translations for English, German, French.  I don't have a Danish translation.

**Comment by leonardpitzu** [comment]
How about adding pictures to it? My local, dumb implementation has a basic if-else approach and shows a picture with the trend. I have the pictures if you want them - both for day and for night.

**Comment by GlennGoddard** [comment]
@leonardpitzu I'm curious regarding the pictures.  It may not make sense in this repository, but I could probably put them to use in my repo for canvas gauges: [https://github.com/GlennGoddard/CanvasGaugeBackgrounds](https://github.com/GlennGoddard/CanvasGaugeBackgrounds)

**Comment by briis** [comment]
Agree with @GlennGoddard. I cannot really see the use case unless we converted it into a Weather Entity, which does not make sense for a forecast like that.

**Comment by GlennGoddard** [comment]
Zambretti is in the Dev build and is under going testing.

**Comment by leonardpitzu** [comment]
Fair enough but in my case i had to do this as it was a family request. Looks like this:
<img width="370" alt="Screenshot 2022-06-29 at 18 25 25" src="https://user-images.githubusercontent.com/4321813/176475823-d2fa5526-b107-41d5-93de-9b3cbc386562.png">

**Comment by GlennGoddard** [comment]
@leonardpitzu Do you have different pictures depending on the weather?

**Comment by leonardpitzu** [comment]
Yes, I have one for each weather condition. Also for each condition a day and night image. I am not in front of my computer right now but i can mail you the images if you want.

**Comment by GlennGoddard** [comment]
You could send them via:
my Repo on GitHub
GlennHA on Home Assistant Forums
GlennGoddard on Tempest Forums

**Comment by GlennGoddard** [comment]
Zambretti will be pushed to Stable branch.  Testing successful in Dev branch.
